### PR TITLE
vstart: Remove old ceph images if new cluster is deployed

### PR DESCRIPTION
**Pull request by varshar16**
Signed-off-by: Varsha Rao <varao@redhat.com>

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
Hm. won't this make things super slow? Often, I run vstart like so: `OSD=0 MDS=0 MON=1 MGR=1 ../src/vstart.sh -n --cephadm --without-dashboard` which spawns a new cluster within 15 secs or so (the majority of the time is wasted by waiting for the restful module). Having to download the image every single time looks excessive. 

otoh I see your point. 

What about asking the registry for the hash of the image and then set cephadm's image config to the sha instead? maybe something like

```bash
digest=curl -s --header "Accept: application/vnd.docker.distribution.manifest.v2+json" \
      "http://docker.io/ceph/daemon-base:latest-master-devel" |
      jq -r '.config.digest'
ceph config set global "docker.io/ceph/daemon-base:$digest"

?

**Comment by varshar16** [comment]
> Hm. won't this make things super slow? Often, I run vstart like so: `OSD=0 MDS=0 MON=1 MGR=1 ../src/vstart.sh -n --cephadm --without-dashboard` which spawns a new cluster within 15 secs or so (the majority of the time is wasted by waiting for the restful module). Having to download the image every single time looks excessive.
> 

Yes, you are right. I also thought about it and I was considering to add additional option to remove old image. But that would introduce another option to vstart.

> otoh I see your point.
> 
> What about asking the registry for the hash of the image and then set cephadm's image config to the sha instead? maybe something like
> 
> ```shell
> digest=curl -s --header "Accept: application/vnd.docker.distribution.manifest.v2+json" \
>       "http://docker.io/ceph/daemon-base:latest-master-devel" |
>       jq -r '.config.digest'
> ceph config set global "docker.io/ceph/daemon-base:$digest"
> 
> ?
> ```

This is a good option. I will update the patch.

**Comment by varshar16** [comment]
@sebastian-philipp Using global does not set the container image for all the services.

```
diff --git a/src/vstart.sh b/src/vstart.sh
index 8b6a84c205..51515a7b5b 100755
--- a/src/vstart.sh
+++ b/src/vstart.sh
@@ -972,9 +972,8 @@ EOF
     if [ "$cephadm" -eq 1 ]; then
         debug echo Enabling cephadm orchestrator
        if [ "$new" -eq 1 ]; then
-               DEFAULT_IMAGE="docker.io/ceph/daemon-base:latest-master-devel"
-               podman image rm $DEFAULT_IMAGE 2>/dev/null || true
-               docker image rm $DEFAULT_IMAGE 2>/dev/null || true
+               digest=$(curl -s https://registry.hub.docker.com/v2/repositories/ceph/daemon-base/tags/latest-master-devel  | jq -r '.images[].digest')
+               ceph_adm config set global container_image "docker.io/ceph/daemon-base@$digest"
        fi
         ceph_adm config-key set mgr/cephadm/ssh_identity_key -i ~/.ssh/id_rsa
         ceph_adm config-key set mgr/cephadm/ssh_identity_pub -i ~/.ssh/id_rsa.pub
```

```
[root@varsha build]# ../src/cephadm/cephadm ls
[
    {
        "style": "cephadm:v1",
        "name": "mds.a.varsha.icblyu",
        "fsid": "ad0a3d1a-7f92-45b4-ac3d-5e6894691b30",
        "systemd_unit": "ceph-ad0a3d1a-7f92-45b4-ac3d-5e6894691b30@mds.a.varsha.icblyu",
        "enabled": true,
        "state": "unknown",
        "container_id": null,
        "container_image_name": "docker.io/ceph/daemon-base@sha256:14ac3fa6c54439119f729a066a212bff09f54213dc53b64f396df719222d5810",
        "container_image_id": null,
        "version": null,
        "started": null,
        "created": "2020-07-14T12:54:55.404653",
        "deployed": "2020-07-14T12:54:54.649640",
        "configured": "2020-07-14T12:54:55.404653"
    },
    {
        "style": "cephadm:v1",
        "name": "nfs.ganesha-vstart.varsha",
        "fsid": "ad0a3d1a-7f92-45b4-ac3d-5e6894691b30",
        "systemd_unit": "ceph-ad0a3d1a-7f92-45b4-ac3d-5e6894691b30@nfs.ganesha-vstart.varsha",
        "enabled": true,
        "state": "running",
        "container_id": "e70b0dee85192f93f4f97162ed7aa30403e77f907c44d158fd9928f4dd4c557b",
        "container_image_name": "docker.io/ceph/daemon-base:latest-master-devel",
        "container_image_id": "549703294cfc467fa7223f68a06b40867a861bdd710b1e7f89edbb6c5810e3f4",
        "version": "3.3",
        "started": "2020-07-14T12:54:59.104801",
        "created": "2020-07-14T12:54:58.526707",
        "deployed": "2020-07-14T12:54:57.767694",
        "configured": "2020-07-14T12:54:58.526707"
    },
    {
        "style": "cephadm:v1",
        "name": "crash.varsha",
        "fsid": "ad0a3d1a-7f92-45b4-ac3d-5e6894691b30",
        "systemd_unit": "ceph-ad0a3d1a-7f92-45b4-ac3d-5e6894691b30@crash.varsha",
        "enabled": true,
        "state": "running",
        "container_id": "3c16bbbbe8150058a6e06ac8359b3f54998b5cdad754871315391d561308e293",
        "container_image_name": "docker.io/ceph/daemon-base:latest-master-devel",
        "container_image_id": "549703294cfc467fa7223f68a06b40867a861bdd710b1e7f89edbb6c5810e3f4",
        "version": "16.0.0-3423-g1904df80be",
        "started": "2020-07-14T12:54:38.166579",
        "created": "2020-07-14T12:54:37.981353",
        "deployed": "2020-07-14T12:54:36.965336",
        "configured": "2020-07-14T12:54:37.981353"
    }
]
```

**Comment by sebastian-philipp** [comment]
interesting. the only explanation I have is that the deployment of ceph-crash was initiated _before_ vstart did set the setting. Try setting container_image before enabling mgr/cephadm.

**Comment by varshar16** [comment]
> interesting. the only explanation I have is that the deployment of ceph-crash was initiated _before_ vstart did set the setting. Try setting container_image before enabling mgr/cephadm.

Here container_image is set before deploying ceph-crash and right image is passed. 

```
2020-07-14T21:25:23.383+0530 7ff7b3c87700  0 [cephadm DEBUG root] osd container image docker.io/ceph/daemon-base@sha256:14ac3fa6c54439119f729a066a212bff09f54213dc53b64f396df719222d5810
2020-07-14T21:25:23.383+0530 7ff7b3c87700  0 [cephadm DEBUG root] args: --image docker.io/ceph/daemon-base@sha256:14ac3fa6c54439119f729a066a212bff09f54213dc53b64f396df719222d5810 ceph-volume --fsid 456cb627-fd5f-4244-81a1-96ffd43b5071 -- inventory --format=json
2020-07-14T21:25:25.259+0530 7ff7b3c87700  0 [cephadm DEBUG root] mon container image docker.io/ceph/daemon-base@sha256:14ac3fa6c54439119f729a066a212bff09f54213dc53b64f396df719222d5810
2020-07-14T21:25:25.259+0530 7ff7b3c87700  0 [cephadm DEBUG root] args: --image docker.io/ceph/daemon-base@sha256:14ac3fa6c54439119f729a066a212bff09f54213dc53b64f396df719222d5810 list-networks
2020-07-14T21:25:25.580+0530 7ff7ad17a700  0 [cephadm DEBUG root] crash.varsha container image docker.io/ceph/daemon-base@sha256:14ac3fa6c54439119f729a066a212bff09f54213dc53b64f396df719222d5810
2020-07-14T21:25:25.580+0530 7ff7ad17a700  0 [cephadm DEBUG root] args: --image docker.io/ceph/daemon-base@sha256:14ac3fa6c54439119f729a066a212bff09f54213dc53b64f396df719222d5810 deploy --fsid 456cb627-fd5f-4244-81a1-96ffd43b5071 --name crash.varsha --config-json - --allow-ptrace

2020-07-14T21:25:43.128+0530 7ff7ad17a700  0 [cephadm DEBUG root] mds.a.varsha.wrkarf container image docker.io/ceph/daemon-base@sha256:14ac3fa6c54439119f729a066a212bff09f54213dc53b64f396df719222d5810
2020-07-14T21:25:43.128+0530 7ff7ad17a700  0 [cephadm DEBUG root] args: --image docker.io/ceph/daemon-base@sha256:14ac3fa6c54439119f729a066a212bff09f54213dc53b64f396df719222d5810 deploy --fsid 456cb627-fd5f-4244-81a1-96ffd43b5071 --name mds.a.varsha.wrkarf --config-json - --allow-ptrace

2020-07-14T21:25:46.244+0530 7ff7ad17a700  0 [cephadm DEBUG root] nfs.ganesha-vstart.varsha container image docker.io/ceph/daemon-base@sha256:14ac3fa6c54439119f729a066a212bff09f54213dc53b64f396df719222d5810
2020-07-14T21:25:46.244+0530 7ff7ad17a700  0 [cephadm DEBUG root] args: --image docker.io/ceph/daemon-base@sha256:14ac3fa6c54439119f729a066a212bff09f54213dc53b64f396df719222d5810 deploy --fsid 456cb627-fd5f-4244-81a1-96ffd43b5071 --name nfs.ganesha-vstart.varsha --config-json - --allow-ptrace
```

**Comment by varshar16** [comment]
Interestingly when I tried again after few times it worked.

**Comment by varshar16** [comment]
jenkins test dashboard backend

File: src/vstart.sh [hunk 1]
```diff
@@ -971,6 +971,12 @@ EOF
 
     if [ "$cephadm" -eq 1 ]; then
         debug echo Enabling cephadm orchestrator
+	if [ "$new" -eq 1 ]; then
+		digest=$(curl -s \
+		https://registry.hub.docker.com/v2/repositories/ceph/daemon-base/tags/latest-master-devel \
+		| jq -r '.images[].digest')
+		ceph_adm config set global container_image "docker.io/ceph/daemon-base@$digest"
```
Comment (sebastian-philipp) [inline src/vstart.sh:978]: ok this works for now. 

I wonder, if we need this in code in cephadm itself at some point. anyway. lgtm

**Pull request merged**